### PR TITLE
🏗 Remove almost all `gulp` streaming from the compilation pipeline

### PR DESCRIPTION
**Pull request by rsimha**
This is another in a series of PRs that modernize our development tasks.

**PR highlights:**
- Rewrite `css/index.js` and `dist.js` without file streaming
- Make all file I/O operations asynchronous
- Some opportunistic clean up (clean up comments / JSDoc, move `copyCss()` to `css/`, etc.)
- Delete `gulp-file` and `gulp-watch` from `package.json`

I've verified that the `build/` and `dist/` output directories (including 3p and tools) are byte-for-byte identical with the same file structure before and after this PR.

After this, the last remaining item is to get rid of all the file streaming in `build-system/compile/compile.js`.

Partial fix for #32585

### CODE REVIEW

**Comment by rsimha** [comment]
I diffed the `build` and `dist` output dirs, they’re unchanged by this PR.

File: build-system/tasks/dist.js [hunk 1]
```diff
@@ -205,23 +201,20 @@ function buildLoginDone(version) {
...
-      const tempBuildDir = `build/all/amp-web-push-${version}/`;
-      const builtName = fileName + '.js';
-      const minifiedName = maybeToEsmName(fileName + '.js');
-      const p = compileJs('./' + tempBuildDir, builtName, './' + distDir, {
+  for (const version of WEB_PUSH_PUBLISHER_VERSIONS) {
+    for (const fileName of WEB_PUSH_PUBLISHER_FILES) {
+      const tempBuildDir = `build/all/amp-web-push-${version}`;
+      const builtName = `${fileName}.js`;
+      const minifiedName = maybeToEsmName(builtName);
+      await compileJs(`./${tempBuildDir}`, builtName, `./${distDir}`, {
```
Comment (samouri) [inline build-system/tasks/dist.js:209]: Why did you change the pattern here? Isn't it slower to waterfall the compilations instead of parallelizing via Promise.all?

File: build-system/tasks/dist.js [hunk 2]
```diff
@@ -231,174 +224,135 @@ async function prebuild() {
...
-      .src('build/css/amp-*.css', {base: 'build/css/'})
-      .pipe(gulp.dest('dist/v0'))
-  ).then(() => {
-    endBuildStep('Copied', 'build/css/*.css to dist/v0/*.css', startTime);
-  });
-}
-
 /**
  * Copies parsers from the build folder to the dist folder
- * @return {!Promise}
```
Comment (samouri) [inline build-system/tasks/dist.js:256]: same nit: async funcs return promises

File: build-system/tasks/dist.js [hunk 3]
```diff
@@ -231,174 +224,135 @@ async function prebuild() {
...
-      promises.push(promise);
-    });
-  });
-  return Promise.all(promises);
+      const js = await fs.readFile(`${srcPath}/${fileName}.js`, 'utf8');
+      const builtName = `${fileName}.js`;
+      await fs.writeFile(`${destPath}/${builtName}`, js);
+      const jsFiles = globby.sync(`${srcPath}/*.js`);
+      for (const jsFile of jsFiles) {
+        await fs.copy(jsFile, `${destPath}/${path.basename(jsFile)}`);
```
Comment (samouri) [inline build-system/tasks/dist.js:252]: same question re. parallelization vs. waterfalling

File: build-system/tasks/dist.js [hunk 4]
```diff
@@ -231,174 +224,135 @@ async function prebuild() {
...
-  return toPromise(
-    gulp
-      .src(path + '/*.js', {base: path})
-      .pipe(file(builtName, js))
-      .pipe(gulp.dest(buildDir))
-  );
+  const js = await fs.readFile(jsPath, 'utf8');
+  const builtName = `amp-login-done-${version}.max.js`;
+  await fs.ensureDir(buildDir);
+  await fs.writeFile(`${buildDir}/${builtName}`, js);
```
Comment (samouri) [inline build-system/tasks/dist.js:351]: tip: i recently learned about `fs.outputFile`, which is the same as ensureDir + writeFile combined.

**Review by samouri — COMMENTED** [review]
You're on fire! 🔥

File: build-system/tasks/css/index.js [hunk 5]
```diff
@@ -15,27 +15,21 @@
...
 const {buildExtensions} = require('../extension-helpers');
+const {endBuildStep, watchDebounceDelay} = require('../helpers');
 const {jsifyCssAsync} = require('./jsify-css');
 const {maybeUpdatePackages} = require('../update-packages');
-const {watch} = require('gulp');
+const {watch} = require('chokidar');
 
 /**
  * Entry point for 'gulp css'
- * @return {!Promise}
```
[the remark this answers is not in the document]
Comment (rsimha) [reply build-system/tasks/css/index.js:34]: This is true, but from what I've seen, our `eslint` JSDoc enforcer requires that a `@return` annotation be added only for functions that actually contain a `return` statement. Functions that are `async` return an implicit Promise, and I think it would be superfluous to document each of them with `@return {!Promise}`.

File: build-system/tasks/dist.js (hunk 1 above)
Comment (rsimha) [reply build-system/tasks/dist.js:209]: I did think of this, and good of you to flag it! I switched the pattern primarily for better readability and predictability of order, and because performance is actually better here (because we're not using vinyl fs). Predictable order helps when we enable debugging for compilation (`gulp dist --debug`).

Here's the reference I used to research this: https://medium.com/dailyjs/the-pitfalls-of-async-await-in-array-loops-cf9cf713bfeb

File: build-system/tasks/dist.js (hunk 2 above)
Comment (rsimha) [reply build-system/tasks/dist.js:256]: Same answer as above. We've gotten rid of `@return` annotations when we've gotten rid of explicit `return` statements in several other places in the codebase.

File: build-system/tasks/dist.js (hunk 3 above)
Comment (rsimha) [reply build-system/tasks/dist.js:252]: Same reason as above. One thing to add is that for cases when there genuinely is reason to parallelize (e.g. building 150 extensions or dozens of JS files at once), we do indeed use `Promise.all`.

File: build-system/tasks/dist.js (hunk 4 above)
Comment (rsimha) [reply build-system/tasks/dist.js:351]: Nice tip! I'm sure we can replace all the `fs.ensureDir` + `fs.writeFile` combos in this file and other files, but we'll still need to figure out what to do with the instances that use `fs.copy` in a loop, for example.

I'm planning on doing further cleanup once I tackle the file streaming in `compile.js`, in case it allows for the simplification of some of this code. Okay to punt on this until then?

**Review by rsimha — COMMENTED** [review]
Nice review comments! This time, you're getting a bunch of replies with no accompanying code changes. Lemme know if you think there's any code that _must_ be changed here.

File: build-system/tasks/css/index.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (samouri) [reply build-system/tasks/css/index.js:34]: Noted! If we ever switch to TS though, it won't be as forgiving

**Review by samouri — APPROVED** [review]
Thanks for addressing all my questions

File: build-system/tasks/css/index.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (rsimha) [reply build-system/tasks/css/index.js:34]: Interesting point. Will keep this in mind if / when we start enforcing TS-style annotations for our build system codebase.

File: build-system/tasks/css/index.js [hunk 6]
```diff
@@ -84,80 +78,77 @@ const cssEntryPoints = [
...
 ];
 
+/**
+ * Copies the css from the build folder to the dist folder
+ */
+async function copyCss() {
+  const startTime = Date.now();
+  for (const {outCss} of cssEntryPoints) {
+    await fs.copy(`build/css/${outCss}`, `dist/${outCss}`);
+  }
```
Comment (jridgewell) [inline build-system/tasks/css/index.js:88]: Nit: parallelize for speeeeeeed
```suggestion
  await Promise.all(cssEntryPoints.map(({outCss}) => {
    return fs.copy(`build/css/${outCss}`, `dist/${outCss}`);
  }));
```

File: build-system/tasks/css/index.js [hunk 7]
```diff
@@ -84,80 +78,77 @@ const cssEntryPoints = [
...
+  const cssFiles = globby.sync('build/css/amp-*.css');
+  for (const cssFile of cssFiles) {
```
Comment (jridgewell) [inline build-system/tasks/css/index.js:90]: Ditto.

**Review by jridgewell — APPROVED** [review]

File: build-system/tasks/css/index.js (hunk 6 above)
Comment (samouri) [reply build-system/tasks/css/index.js:88]: lol, i had the same comment 🚀

Comment (rsimha) [reply build-system/tasks/css/index.js:88]: Whaddayaknow, I cannot possibly say no to good ideas from both of you, so here I am sitting, addressing a bunch of other comments from Jake's first review that I initially thought I'd punt on 😃

File: build-system/tasks/dist.js (hunk 4 above)
Comment (rsimha) [reply build-system/tasks/dist.js]: Did this and removed every useless call to `ensureDir`.

File: build-system/tasks/dist.js (hunk 1 above)
Comment (rsimha) [reply build-system/tasks/dist.js:209]: Update: I've parallelized all file copies that no longer need to be serial. There are one or two instances like compiling two related parts of the same module that still happen one after the other. Nice feedback!

File: build-system/tasks/css/index.js (hunk 6 above)
Comment (rsimha) [reply build-system/tasks/css/index.js:89]: Parallelizing the top-level CSS copying results in this race caught by `gulp check-sourcemaps` with extra logging: https://app.circleci.com/pipelines/github/ampproject/amphtml/2953/workflows/93cf8701-336b-4a00-b4cd-332a8076f574/jobs/34479

I've reverted it for now, can optimize in a future refactor.
